### ISSUE TITLE
bls_api() returning corrupted data

### ISSUE THREAD

**Issue by tomhopper** [keberwein/blscrapeR#21]
`bls_api()` is returning corrupted data. A call to a date range that includes the latest record of one of the data series results in a warning "number of columns of result is not a multiple of vector length (arg 2)." When this happens, the first and most recent record appears to have an extra column, _latest_, that is missing in subsequent records. `bls_api()` seems to be handling this by filling the data from subsequent rows and then padding the last column with data from the first column, resulting in corruption of the data frame.

The columns for the first row appear to be

```
year    period    periodName    latest    value    footnotes    seriesID
```

where _latest_ is only present for the first/most recent record, and contains the value `TRUE`.

In subsequent rows, the column _latest_ appears to missing, but the data is filled in using values from the columns:

```
year    period    periodName    value    footnotes    seriesID    year
````

Below are four example codes. The first two show the results of the call to bls_api() directly, and the second two show the result after tidyr::unnest() to convert _list_ columns to vectors.

The first and third examples are successful, pulling data as expected for years 1980 through 2000. The second and fourth both show the problem, pulling data for years 2000 through 2018. The latest record that actually exists (at time of this writing) is for period M12 of 2017, and when pulled through the BLS site interface at [https://data.bls.gov/cgi-bin/srgate](https://data.bls.gov/cgi-bin/srgate) is reported as preliminary. i.e., the row on the website looks like

```
LAUMT263474000000003 | 2017 | M12 | 5.3(P)
```

```
> bls_api(seriesid = "LAUMT263474000000003", startyear = 1980, endyear = 2000, registrationKey = Sys.getenv("BLS_KEY"))
REQUEST_SUCCEEDED
# A tibble: 120 x 6
    year period    periodName value footnotes seriesID 
   <dbl> <list>    <list>     <dbl> <list>    <list>   
 1  1999 <chr [1]> <chr [1]>   3.60 <chr [1]> <chr [1]>
 2  1999 <chr [1]> <chr [1]>   3.80 <chr [1]> <chr [1]>
 3  1999 <chr [1]> <chr [1]>   3.70 <chr [1]> <chr [1]>
 4  1999 <chr [1]> <chr [1]>   3.70 <chr [1]> <chr [1]>
 5  1999 <chr [1]> <chr [1]>   3.80 <chr [1]> <chr [1]>
 6  1999 <chr [1]> <chr [1]>   4.50 <chr [1]> <chr [1]>
 7  1999 <chr [1]> <chr [1]>   5.10 <chr [1]> <chr [1]>
 8  1999 <chr [1]> <chr [1]>   4.30 <chr [1]> <chr [1]>
 9  1999 <chr [1]> <chr [1]>   4.20 <chr [1]> <chr [1]>
10  1999 <chr [1]> <chr [1]>   4.20 <chr [1]> <chr [1]>
# ... with 110 more rows
```
```
> bls_api(seriesid = "LAUMT263474000000003", startyear = 2000, endyear = 2018, registrationKey = Sys.getenv("BLS_KEY"))
REQUEST_SUCCEEDED
# A tibble: 216 x 7
    year period    periodName latest    value footnotes seriesID 
   <dbl> <list>    <list>     <list>    <dbl> <list>    <list>   
 1  2017 <chr [1]> <chr [1]>  <chr [1]>  5.30 <chr [1]> <chr [1]>
 2  2017 <chr [1]> <chr [1]>  <chr [1]> NA    <chr [1]> <chr [1]>
 3  2017 <chr [1]> <chr [1]>  <chr [1]> NA    <chr [1]> <chr [1]>
 4  2017 <chr [1]> <chr [1]>  <chr [1]> NA    <chr [1]> <chr [1]>
 5  2017 <chr [1]> <chr [1]>  <chr [1]> NA    <chr [1]> <chr [1]>
 6  2017 <chr [1]> <chr [1]>  <chr [1]> NA    <chr [1]> <chr [1]>
 7  2017 <chr [1]> <chr [1]>  <chr [1]> NA    <chr [1]> <chr [1]>
 8  2017 <chr [1]> <chr [1]>  <chr [1]> NA    <chr [1]> <chr [1]>
 9  2017 <chr [1]> <chr [1]>  <chr [1]> NA    <chr [1]> <chr [1]>
10  2017 <chr [1]> <chr [1]>  <chr [1]> NA    <chr [1]> <chr [1]>
# ... with 206 more rows
Warning message:
In rbind(list(year = "2017", period = "M12", periodName = "December",  :
  number of columns of result is not a multiple of vector length (arg 2)
```
## With tidyr::unbind()
```
> bls_api(seriesid = "LAUMT263474000000003", startyear = 1980, endyear = 2000, registrationKey = Sys.getenv("BLS_KEY")) %>% 
+   tidyr::unnest()
REQUEST_SUCCEEDED
# A tibble: 120 x 6
    year value period periodName footnotes seriesID            
   <dbl> <dbl> <chr>  <chr>      <chr>     <chr>               
 1  1999  3.60 M12    December   ""        LAUMT263474000000003
 2  1999  3.80 M11    November   ""        LAUMT263474000000003
 3  1999  3.70 M10    October    ""        LAUMT263474000000003
 4  1999  3.70 M09    September  ""        LAUMT263474000000003
 5  1999  3.80 M08    August     ""        LAUMT263474000000003
 6  1999  4.50 M07    July       ""        LAUMT263474000000003
 7  1999  5.10 M06    June       ""        LAUMT263474000000003
 8  1999  4.30 M05    May        ""        LAUMT263474000000003
 9  1999  4.20 M04    April      ""        LAUMT263474000000003
10  1999  4.20 M03    March      ""        LAUMT263474000000003
# ... with 110 more rows
```
```
> bls_api(seriesid = "LAUMT263474000000003", startyear = 2000, endyear = 2018, registrationKey = Sys.getenv("BLS_KEY")) %>% 
+   tidyr::unnest()
REQUEST_SUCCEEDED
# A tibble: 216 x 7
    year value period periodName latest footnotes            seriesID            
   <dbl> <dbl> <chr>  <chr>      <chr>  <chr>                <chr>               
 1  2017  5.30 M12    December   true   P Preliminary.       LAUMT263474000000003
 2  2017 NA    M11    November   4.9    LAUMT263474000000003 2017                
 3  2017 NA    M10    October    5.2    LAUMT263474000000003 2017                
 4  2017 NA    M09    September  5.3    LAUMT263474000000003 2017                
 5  2017 NA    M08    August     5.8    LAUMT263474000000003 2017                
 6  2017 NA    M07    July       6.2    LAUMT263474000000003 2017                
 7  2017 NA    M06    June       4.8    LAUMT263474000000003 2017                
 8  2017 NA    M05    May        4.4    LAUMT263474000000003 2017                
 9  2017 NA    M04    April      3.9    LAUMT263474000000003 2017                
10  2017 NA    M03    March      4.8    LAUMT263474000000003 2017                
# ... with 206 more rows
Warning message:
In rbind(list(year = "2017", period = "M12", periodName = "December",  :
  number of columns of result is not a multiple of vector length (arg 2)
```

**Comment by tomhopper** [comment]
There is a work-around, at least when we know the problem exists:

```
bls_api(seriesid = "LAUMT263474000000003", startyear = 2000, endyear = 2018, registrationKey = Sys.getenv("BLS_KEY")) %>% 
  filter(row_number() != 1) %>% 
  tidyr::unnest()  %>% 
  select(-seriesID) %>% 
  rename('seriesID' = "footnotes",
         'footnotes' = "value",
         'value' = "latest") %>% 
  mutate(value = as.numeric(value))
```

I have not figured out, yet, how to programmatically check for the problem.

**Comment by keberwein** [comment]
Ah, thanks for pointing this out. Looks like they made a change to the API. In the past they've emailed developers with any changes--guess not this time.

I've seen the "latest" column in other data sets. It only occurs on with the most current month in data set.

I'm pretty sure I know how to fix the issue, but I don't think I'll have time to get around to it until early next week.

I'll keep this issue open until I have deployed the fix and pushed a new version to CRAN.

Thanks again for the head's up!

**Comment by keberwein** [comment]
This issue has been resolved in the current development version 3.1.2. I replaced the old `do.call(map)` method with `purrr::map_dfr`, which allows the resulting tibble to adjust automatically to the data.

API queries up to the current month now have an extra column with the 'latest=TRUE' status.

The dev. version can be installed directly from GitHub. I will be pushing it to CRAN in the next week or so; after I finish some more testing.

`devtools::install_github("keberwein/blscrapeR")`

```
test1 <- bls_api(seriesid = "LAUMT263474000000003", startyear = 1999, endyear = 2000,
 registrationKey = Sys.getenv("BLS_KEY"))
```

```
test2 <- bls_api(seriesid = "LAUMT263474000000003", startyear = 2017, endyear = 2018, 
registrationKey = Sys.getenv("BLS_KEY"))
```
